src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,456 @@
+		}
+	}
+
+	// Assign weights with descending priority (most weight to first samples)
+	const size_t max_weight = 10;
+	seek_sample_weights_.resize(seek_samples_.size());
+	for (size_t i = 0; i < seek_samples_.size(); ++i) {
+		seek_sample_weights_[i] = static_cast<size_t>(max_weight - i);
+	}
+}
+
+size_t DiskNoiseDevice::ChooseWeightedSeekIndex() const
+{
+	size_t total_weight = 0;
+	for (size_t i = 0; i < seek_samples_.size(); ++i) {
+		if (!seek_samples_[i].empty()) {
+			total_weight += seek_sample_weights_[i];
+		}
+	}
+
+	if (total_weight == 0) {
+		return 0;
+	}
+
+	const size_t r = static_cast<size_t>(rand()) % total_weight;
+	size_t sum     = 0;
+	for (size_t i = 0; i < seek_samples_.size(); ++i) {
+		if (seek_samples_[i].empty()) {
+			continue;
+		}
+
+		sum += seek_sample_weights_[i];
+		if (r < sum) {
+			return i;
+		}
+	}
+	return 0;
+}
+
+DiskNoiseDevice::DiskNoiseDevice(const bool enable_disk_noise,
+                                 const std::string& spin_up_sample_path,
+                                 const std::string& spin_sample_path,
+                                 const std::vector<std::string>& seek_sample_paths,
+                                 const std::string& spin_channel_name,
+                                 const std::string& seek_channel_name,
+                                 const unsigned int& spin_volume,
+                                 const unsigned int& seek_volume)
+        : enable_disk_noise_(),
+          spin_up_sample_(),
+          spin_sample_(),
+          current_seek_sample_(),
+          seek_sample_weights_(),
+          seek_samples_(),
+          spin_pos_(0),
+          seek_pos_(0),
+          last_activity_(0),
+          spin_channel_(nullptr),
+          seek_channel_(nullptr)
+{
+	enable_disk_noise_ = enable_disk_noise;
+	if (!enable_disk_noise_) {
+		LOG_INFO("DiskNoiseDevice: Disk noise emulation disabled");
+		return;
+	}
+
+	LOG_DEBUG("DiskNoiseDevice: Spin sample volume: %u", spin_volume);
+	LOG_DEBUG("DiskNoiseDevice: Seek sample volume: %u", seek_volume);
+
+	spin_volume_ = spin_volume;
+	seek_volume_ = seek_volume;
+	LoadSample(spin_up_sample_path, spin_up_sample_);
+	LoadSample(spin_sample_path, spin_sample_);
+	LoadSeekSamples(seek_sample_paths);
+
+	using namespace std::placeholders;
+
+	MIXER_LockMixerThread();
+
+	auto spin_callback = std::bind(&DiskNoiseDevice::AudioCallbackSpin,
+	                               this,
+	                               std::placeholders::_1);
+
+	auto seek_callback = std::bind(&DiskNoiseDevice::AudioCallbackSeek,
+	                               this,
+	                               std::placeholders::_1);
+
+	spin_channel_ = MIXER_AddChannel(spin_callback,
+	                                 44100,
+	                                 spin_channel_name.c_str(),
+	                                 {ChannelFeature::Stereo});
+	seek_channel_ = MIXER_AddChannel(seek_callback,
+	                                 44100,
```

Comment:
This should be a constant, then you can use the same constant to verify the WAV files are 44.1 kHz, otherwise log an error and disable disk noises.